### PR TITLE
Fix: Websocket upgrade when used with Http.collectM

### LINKED ISSUE #623: Not able to create websocket multiple connections when Response is ZIO
**Describe the bug**
```scala
  private val socketApp = SocketApp.message(socket)
  private val webSocketApp = Http.collectM[Request] { case Method.GET -> !! / "subscriptions" =>
    ZIO(Response.socket(socketApp))
  }
  ```
  using the above code not able to make more than 5 connections

**To Reproduce**
1.  Create a websocketApp using below code
```scala
  private val socketApp = SocketApp.message(socket)
  private val webSocketApp = Http.collectM[Request] { case Method.GET -> !! / "subscriptions" =>
    ZIO(Response.socket(socketApp))
  }
  ```
2. Start server and make multiple websocket request. The request go in pending after couple of successful request.

**Expected behavior**
The socket request should upgrade to 101, till the time ports and file descriptors are available for client

### LINKED ISSUE #640: Add Integration tests for websocket
**What is not easy to do right now?**
Right now, there is no way to test WebSocket

**Describe the solution you'd like**
We can add functionality to the existing client to create WebSocket connections, meanwhile, we can use any other websocket client to implement integration test

### PR DESCRIPTION
**Pull request by amitksingh1490**
To resolve #623

### CODE REVIEW

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 431355.36**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 438593.67**

File: zio-http/src/test/scala/zhttp/service/ServerSpec.scala [hunk 1]
```diff
@@ -181,7 +180,7 @@ object ServerSpec extends HttpRunnableSpec(8088) {
   } yield (data.mkString(""), content)
 
   private val env = EventLoopGroup
-    .nio() ++ ChannelFactory.nio ++ ServerChannelFactory.nio ++ AppCollection.live
+    .nio(10) ++ ChannelFactory.nio ++ ServerChannelFactory.nio ++ AppCollection.live
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/service/ServerSpec.scala]: why is this required?

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 433440.28**

File: zio-http/src/test/scala/zhttp/service/ServerSpec.scala [hunk 2]
```diff
@@ -194,8 +193,8 @@ object ServerSpec extends HttpRunnableSpec(8088) {
     }
 
   private val socketApp = SocketApp.message(socket) ++ SocketApp.protocol(SocketProtocol.handshakeTimeout(100 millis))
-  private val webSocketApp = Http.collect[Request] { case Method.GET -> !! / "subscriptions" =>
-    Response.socket(socketApp)
+  private val webSocketApp = Http.collectM[Request] { case Method.GET -> !! / "subscriptions" =>
+    ZIO(Response.socket(socketApp))
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/service/ServerSpec.scala]: use `Http.fromEffect` instead.

File: zio-http/src/test/scala/zhttp/service/ServerSpec.scala (hunk 1 above)
Comment (amitksingh1490) [reply zio-http/src/test/scala/zhttp/service/ServerSpec.scala]: .nio(10) is Not Required.

File: project/Dependencies.scala [hunk 3]
```diff
@@ -6,14 +6,16 @@ object Dependencies {
...
   val `scala-compact-collection` = "org.scala-lang.modules" %% "scala-collection-compat" % ScalaCompactCollectionVersion
   val netty                      = "io.netty"                % "netty-all"               % NettyVersion
   val `jwt-core`                 = "com.github.jwt-scala"   %% "jwt-core"                % JwtCoreVersion
-  val zio                        = "dev.zio"                %% "zio"                     % ZioVersion
-  val `zio-streams`              = "dev.zio"                %% "zio-streams"             % ZioVersion
-  val `zio-test`                 = "dev.zio"                %% "zio-test"                % ZioVersion % "test"
-  val `zio-test-sbt`             = "dev.zio"                %% "zio-test-sbt"            % ZioVersion % "test"
-  val `netty-incubator`          =
+  val sttp              = "com.softwaremill.sttp.client3" %% "core"                          % sttpVersion
+  val sttpzio           = "com.softwaremill.sttp.client3" %% "async-http-client-backend-zio" % sttpVersion
```
Comment (tusharmath) [inline project/Dependencies.scala]: `sttp` should only be a test dependency.

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 426459.94**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 428397.20**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 440040.60**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 450969.18**

File: zio-http/src/test/scala/zhttp/service/ServerSpec.scala (hunk 2 above)
Comment (amitksingh1490) [reply zio-http/src/test/scala/zhttp/service/ServerSpec.scala]: This can't be done since we have another app composed after this if I use `Http.fromEffect`, all request will match this app and none will go to the next app

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 438432.59**

File: project/Dependencies.scala (hunk 3 above)
Comment (amitksingh1490) [reply project/Dependencies.scala]: Done

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 429257.48**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 442581.48**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 431599.60**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 433859.59**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 438429.97**

File: zio-http/src/test/scala/zhttp/service/ServerSpec.scala [hunk 4]
```diff
@@ -168,10 +178,16 @@ object ServerSpec extends HttpRunnableSpec(8088) {
...
+  private val socket    =
+    Socket.collect[WebSocketFrame] { case WebSocketFrame.Text("FOO") =>
+      ZStream.succeed(WebSocketFrame.text("BAR"))
+    }
+  private val socketApp = SocketApp.message(socket) ++ SocketApp.protocol(SocketProtocol.handshakeTimeout(100 millis))
   private val staticApp = Http.collectM[Request] {
     case Method.GET -> !! / "success"       => ZIO.succeed(Response.ok)
     case Method.GET -> !! / "failure"       => ZIO.fail(new RuntimeException("FAILURE"))
     case Method.GET -> !! / "get%2Fsuccess" => ZIO.succeed(Response.ok)
+    case Method.GET -> !! / "subscriptions" => ZIO(Response.socket(socketApp))
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/service/ServerSpec.scala]: Don't change the `staticApp`, it's only here for legacy reasons.

File: zio-http/src/test/scala/zhttp/service/ServerSpec.scala [hunk 5]
```diff
@@ -157,8 +161,14 @@ object ServerSpec extends HttpRunnableSpec(8088) {
...
+      } +
+      testM("Multiple Websocket Requests") {
+        val p = (for {
+          res <- send(basicRequest.get(uri"ws://localhost:8088/subscriptions").response(asWebSocketUnsafe))
+        } yield res.code).repeatN(1024)
+        assertM(p)(equalTo(StatusCode(101)))
 
-      }
+      }.provideCustomLayer(AsyncHttpClientZioBackend.layer().orDie)
   }
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/service/ServerSpec.scala]: - Create a separate `WebSocketServerSpec.scala` and move the tests there.
- Ensure no tests changes are required when ZIO-Http Client supports websockets.

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 437531.33**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 427793.97**

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 412473.80**

File: zio-http/src/test/scala/zhttp/internal/HttpRunnableSpec.scala [hunk 6]
```diff
@@ -27,6 +29,13 @@ abstract class HttpRunnableSpec(port: Int) extends DefaultRunnableSpec { self =>
       )
       .map(_.status)
 
+  def websocketStatus(path: Path): ZIO[SttpClient, Throwable, Status] = {
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/internal/HttpRunnableSpec.scala]:
```suggestion
  def websocketStatusCode(path: Path): ZIO[SttpClient, Throwable, Int] = {
```

File: zio-http/src/test/scala/zhttp/service/WebsocketServerSpec.scala [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+  }
+
+  private val env =
+    EventLoopGroup.nio() ++ ServerChannelFactory.nio ++ AsyncHttpClientZioBackend.layer().orDie
+
+  private val socket    =
+    Socket.collect[WebSocketFrame] { case WebSocketFrame.Text("FOO") =>
+      ZStream.succeed(WebSocketFrame.text("BAR"))
+    }
+  private val socketApp = SocketApp.message(socket) ++ SocketApp.protocol(SocketProtocol.handshakeTimeout(100 millis))
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/service/WebsocketServerSpec.scala]: don't create a global SocketApp. Use the same technique as `ServerSpec` and keep the HttpApp close to the test. You will have to deploy the HttpApp, get the url and then make the request.

Learn more about  it on from `RunnableHttpAppSyntax.request`

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 434062.78**

File: zio-http/src/test/scala/zhttp/internal/HttpRunnableSpec.scala [hunk 8]
```diff
@@ -50,6 +64,17 @@ abstract class HttpRunnableSpec(port: Int) extends DefaultRunnableSpec { self =>
 
   implicit class RunnableHttpAppSyntax(app: HttpApp[HttpEnv, Throwable]) {
     def deploy: ZIO[HttpAppCollection, Nothing, String] = AppCollection.deploy(app)
+
+    def websocketStatusCode(
+      path: Path = !!,
+      headers: List[Header] = Nil,
+    ): ZIO[SttpClient with HttpAppCollection, Throwable, Status] = (for {
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/internal/HttpRunnableSpec.scala]: Return `Int` not the status.

File: zio-http/src/test/scala/zhttp/service/WebsocketServerSpec.scala [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+  override def spec = suiteM("Server") {
+    app.as(List(websocketSpec)).useNow
+  }.provideCustomLayerShared(env) @@ timeout(30 seconds)
+
+  def websocketSpec = suite("Websocket Server") {
+    suite("connections") {
+      testM("Multiple websocket upgrades") {
+        val socketApp = SocketApp.message(Socket.succeed(WebSocketFrame.text("BAR"))) ++ SocketApp.protocol(
+          SocketProtocol.handshakeTimeout(100 millis),
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/service/WebsocketServerSpec.scala]: why do we need a timeout?

Comment (amitksingh1490) [reply zio-http/src/test/scala/zhttp/service/WebsocketServerSpec.scala]: we can remove it.

**Comment by amitksingh1490** [comment]
## 🚀🚀🚀 Benchmark Results 
 **Concurrency: 256** 
 **Requests/sec: 446254.57**

**Review by tusharmath — APPROVED** [review]
